bin/faucet/src/client.rs

```diff
@@ -0,0 +1,353 @@
+unsafe impl Sync for FaucetClient {}
+
+impl FaucetClient {
+    pub async fn new(config: FaucetConfig) -> Result<Self, FaucetError> {
+        let (rpc_api, root_block_header, root_chain_mmr) =
+            initial_connection(config.clone()).await?;
+
+        let (faucet_account, account_seed, secret) = build_account(config.clone())?;
+        let faucet_account = Rc::new(RefCell::new(faucet_account));
+        let id = faucet_account.borrow().id();
+
+        let data_store = FaucetDataStore::new(
+            faucet_account.clone(),
+            account_seed,
+            root_block_header,
+            root_chain_mmr,
+        );
+        let authenticator = FaucetAuthenticator::new(secret);
+        let executor = TransactionExecutor::new(data_store.clone(), Some(Rc::new(authenticator)));
+
+        let mut rng = thread_rng();
+        let coin_seed: [u64; 4] = rng.gen();
+        let rng = RpoRandomCoin::new(coin_seed.map(Felt::new));
+        Ok(Self { data_store, rpc_api, executor, id, rng })
+    }
+
+    pub fn execute_mint_transaction(
+        &mut self,
+        target_account_id: AccountId,
+        is_private_note: bool,
+        asset_amount: u64,
+    ) -> Result<(ExecutedTransaction, Note), FaucetError> {
+        let asset = FungibleAsset::new(self.id, asset_amount)
+            .map_err(|err| FaucetError::InternalServerError(err.to_string()))?;
+
+        let note_type = if is_private_note {
+            NoteType::OffChain
+        } else {
+            NoteType::Public
+        };
+
+        let output_note =
+            create_p2id_note(self.id, target_account_id, vec![asset.into()], note_type, self.rng)
+                .map_err(|err| FaucetError::InternalServerError(err.to_string()))?;
+
+        let recipient = output_note
+            .recipient()
+            .digest()
+            .iter()
+            .map(|x| x.as_int().to_string())
+            .collect::<Vec<_>>()
+            .join(".");
+
+        let tag = output_note.metadata().tag().inner();
+
+        let script = ProgramAst::parse(
+            &DISTRIBUTE_FUNGIBLE_ASSET_SCRIPT
+                .replace("{recipient}", &recipient)
+                .replace("{note_type}", &Felt::new(note_type as u64).to_string())
+                .replace("{tag}", &Felt::new(tag.into()).to_string())
+                .replace("{amount}", &Felt::new(asset.amount()).to_string()),
+        )
+        .expect("shipped MASM is well-formed");
+
+        let script = self.executor.compile_tx_script(script, vec![], vec![]).map_err(|err| {
+            FaucetError::InternalServerError(format!("Failed to compile script: {}", err))
+        })?;
+
+        let mut transaction_args = TransactionArgs::new(Some(script), None, AdviceMap::new());
+        transaction_args.extend_expected_output_notes(vec![output_note.clone()]);
+
+        self.executor
+            .load_account(self.id)
+            .map_err(|err| FaucetError::InternalServerError(err.to_string()))?;
```

Comment:
The account can probably be loaded once on initialization instead of on every request